azurelinuxagent/common/cgroup.py

```diff
@@ -250,53 +288,103 @@ def get_tracked_metrics(self, **kwargs):
         tracked = []
         cpu_usage = self.get_cpu_usage()
         if cpu_usage >= float(0):
-            tracked.append(MetricValue(MetricsCategory.CPU_CATEGORY, MetricsCounter.PROCESSOR_PERCENT_TIME, self.name, cpu_usage))
+            tracked.append(
+                MetricValue(MetricsCategory.CPU_CATEGORY, MetricsCounter.PROCESSOR_PERCENT_TIME, self.name, cpu_usage))
 
         if 'track_throttled_time' in kwargs and kwargs['track_throttled_time']:
             throttled_time = self.get_cpu_throttled_time()
             if cpu_usage >= float(0) and throttled_time >= float(0):
-                tracked.append(MetricValue(MetricsCategory.CPU_CATEGORY, MetricsCounter.THROTTLED_TIME, self.name, throttled_time))
+                tracked.append(
+                    MetricValue(MetricsCategory.CPU_CATEGORY, MetricsCounter.THROTTLED_TIME, self.name, throttled_time))
 
         return tracked
 
 
 class MemoryCgroup(CGroup):
+
+    def _get_memory_stat_counter(self, counter_name, raise_error_if_counter_not_present=True):
+        try:
+            with open(os.path.join(self.path, 'memory.stat')) as memory_stat:
+                # cat /sys/fs/cgroup/memory/azure.slice/memory.stat
+                # cache 67178496
+                # rss 42340352
+                # rss_huge 6291456
+                # swap 0
+                for line in memory_stat:
+                    re_memory_counter = r'{0}\s+(\d+)'.format(counter_name)
+                    match = re.match(re_memory_counter, line)
+                    if match is not None:
+                        return int(match.groups()[0])
+        except (IOError, OSError) as e:
+            if e.errno == errno.ENOENT:
+                raise
+            raise CGroupsException("Failed to read memory.stat: {0}".format(ustr(e)))
+        except Exception as e:
+            raise CGroupsException("Failed to read memory.stat: {0}".format(ustr(e)))
+
+        if raise_error_if_counter_not_present:
+            raise Exception("Cannot find counter: {0}".format(counter_name))
+        else:
+            return 0
+
     def get_memory_usage(self):
         """
-        Collect memory.usage_in_bytes from the cgroup.
+        Collect RSS+CACHE from memory.stat cgroup.
 
         :return: Memory usage in bytes
         :rtype: int
         """
-        usage = None
         try:
-            usage = self._get_parameters('memory.usage_in_bytes', first_line_only=True)
-        except Exception as e:
-            if isinstance(e, (IOError, OSError)) and e.errno == errno.ENOENT:  # pylint: disable=E1101
+            cache = self._get_memory_stat_counter("cache")
+            rss = self._get_memory_stat_counter("rss")
+            return cache + rss
+        except (IOError, OSError) as e:
```

Comment:
can we avoid the repetition of the exception handling logic in _get_memory_stat_counter, _get_memory_stat_counter and get_swap_memory_usage?